src/NServiceBus.Core/Unicast/StartAndStoppablesRunner.cs

```diff
@@ -0,0 +1,90 @@
+﻿namespace NServiceBus.Unicast
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NServiceBus.Logging;
+
+    class StartAndStoppablesRunner
+    {
+        public StartAndStoppablesRunner(IEnumerable<IWantToRunWhenBusStartsAndStops> wantToRunWhenBusStartsAndStops)
+        {
+            wantToRunWhenBusStartsAndStopses = wantToRunWhenBusStartsAndStops;
+        }
+
+        public Task StartAsync()
+        {
+            var startableTasks = new List<Task>();
+            foreach (var startable in wantToRunWhenBusStartsAndStopses)
+            {
+                var task = startable.StartAsync();
+
+                var startable1 = startable;
+                task.ContinueWith(t =>
+                {
+                    thingsRanAtStartup.Add(startable1);
+                    Log.DebugFormat("Started {0}.", startable1.GetType().AssemblyQualifiedName);
+                }, TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously);
+                task.ContinueWith(t => { Log.Error("Startup task failed to complete.", t.Exception); }, TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously);
+
+                startableTasks.Add(task);
+            }
+
+            return Task.WhenAll(startableTasks.ToArray());
+        }
+
+        public async Task StopAsync()
+        {
+            var stoppables = Interlocked.Exchange(ref thingsRanAtStartup, new ConcurrentBag<IWantToRunWhenBusStartsAndStops>());
+            if (!stoppables.Any())
+            {
+                return;
+            }
+
+            var stoppableTasks = new List<Task>();
+            foreach (var stoppable in stoppables)
+            {
+                try
+                {
+                    var task = stoppable.StopAsync();
+
+                    var stoppable1 = stoppable;
+                    task.ContinueWith(t =>
+                    {
+                        thingsRanAtStartup.Add(stoppable1);
+                        Log.DebugFormat("Stopped {0}.", stoppable1.GetType().AssemblyQualifiedName);
+                    }, TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously)
+                        .Ignore();
+                    task.ContinueWith(t =>
+                    {
+                        Log.Fatal("Startup task failed to stop.", t.Exception);
```

Comment:
We've got a `FeatureStartupTask` which can be stopped as well. Should we be more explicit and clearly state that 
"`stoppable1.GetType().AssemblyQualifiedName` failed to stop" to avoid any confusion with (feature) startup tasks??